### DISCUSSION TITLE
Use remix with wildcard subdomain

### DISCUSSION THREAD

**Discussion by lgraubner** [remix-run/remix#1475] [Q&A]
Hi!

I'm thinking about building an application which uses one subdomain per user. So for example my app has one main domain, where it shows a simple landingpage, register, login etc. If the app is accessed with a subdomain (which is tied to an user) it should show different content, both public and also a authentication required area.

I made some shallow tests and it seems to work just fine. The authenticated part is the same for every user, just with a subdomain as additional identifier. Routing between domain/subdomains happens only via redirect in loader and not client side.

Are there any downsides I'm missing?

**Comment by a-type** [comment]
Hi @lgraubner , would you mind sharing some more details about your explorations? I'm considering using Remix for an app which allows branded subdomains and I'm interested in how you approached it. Is there a difference between the way the app renders on `mainsite.com/` vs `user.mainsite.com/`, and if so, how did you include the logic to differentiate those within the app?

I'm specifically looking to do a subdomain system where `someone.mainsite.com/x` is equivalent to `mainsite.com/someone/x` for any nested `x`. I'm having some trouble wrapping my head around doing that with filesystem routing. It feels like I'll end up having to include two parallel identical route trees in all of my route modules...

```
routes/
|__ $username/
|    |__ x.tsx  <-- matches mainsite.com/someone/x
|__ x.tsx <-- matches someone.mainsite.com/x
```

This doesn't seem practical at all client-side. Really feels like some server middleware is in order.

Of course Remix does seem to expose plenty of opportunities to modify server behavior. I'm going to explore some more and see... `entry.server.tsx` looks promising, in particular.

**Reply by a-type** [reply]
A little experimenting and doing something like this for `/api/index.js` (with a Vercel deployment) seems to work, so far.

```js
const { createRequestHandler } = require("@remix-run/vercel");
const invariant = require("tiny-invariant");

const innerHandler = createRequestHandler({
  build: require("./_build"),
});

invariant(!!process.env.APEX_URL, "APEX_URL is required");
const SUBDOMAIN_MATCH = new RegExp(`^(.+?)\.${process.env.APEX_URL}$`);

/**
 * @param {import('@vercel/node').VercelRequest} request
 * @param {import('@vercel/node').VercelResponse} response
 */
module.exports = (request, response) => {
  const host = request.headers["x-forwarded-host"] || request.headers["host"];
  const result = host.match(SUBDOMAIN_MATCH);
  if (result && result[1] !== "www") {
    const subdomain = result[1];
    request.url = `/${subdomain}${request.url}`;
  }

  return innerHandler(request, response);
};
```

These server-side solutions seem to be platform-specific. I tried getting closer to Remix in `entry.server.tsx` but seemed to run afoul of route-matching logic which would require more digging to get to the bottom of.

**Reply by lgraubner** [reply]
Sure!

First of all, react-router is limited to one domain/subdomain. You can't route between domains client side (with pure JavaScript it works of course). Therefore it's probably best to make any redirects in the loaders.

As remix routing is based on filesystem routing you have to manually distinguish domains and adjust rendered output accordingly. That should work just fine, you can check in your loader if a domain condition is met, load stuff from database and pass it to the frontend. Meta data can adapt accordingly based on that data.

Where you have to be careful is nested routes. If a top level route can have different layouts and data it might be difficult (not impossible) to react accordingly in the leaf routes. I would suggest to keep the route structure flat for those routes.

For your specific case: In a more traditional web app you would probably rewrite the subdomain to the path with Nginx or similar which should work fine. But with remix I don't think that will work. Paths for react-router would be probably screwed up. Your idea to duplicate the site in the filesystem should work fine. You can extract UI, loader, etc. and just wire them for each page.

For me I'm probably trying to avoid paths used by both "parts". So `app.somedomain.com` will be used to access a protected area (with routes such as `/login`, `/dashboard`, etc.). `*.somedomain.com` shows a user specific profile page. I'm redirecting `app.somedomain.com` to `/dashboard` to avoid two different pages in one route. This will probably not work for every usecase and I'm still evaluating if this approach fits my needs.

I'm using some helper methods to determine in which "context" I am and respond accordingly:

```javascript
function getSubdomain(request: Request) {
  let url = new URL(request.url);

  let parts = url.hostname.split('.');
  // invalid
  invariant(parts.length > 1, 'Expected domain parts to be more than 1.');

  if (parts.length === 2) {
    // not a subdomain
    return null;
  }

  return parts[0];
}

function ensureAppContext(request: Request): string | null {
  let subdomain = getSubdomain(request);

  if (subdomain !== 'app') {
    throw new Response('Not found', {
      status: 404,
    });
  }

  return subdomain;
}

export let loader = ({ request }) => {
  ensureAppContext(request)

  // ...
}
```

**Reply by a-type** [reply]
Thanks for the details. Since my idea is to utilize custom subdomains to let users publish content publically, I'd like to keep my options open for nested routes. But perhaps there's a way to organize my app's structure so the routes are nicely separated.

I think if I just give everyone a subdomain by default, I could just not have the `/prefix/x` pattern and only support subdomains. Trying to do both might be the source of unreasonable complexity here.

**Reply by lgraubner** [reply]
I think it all boils down to there is no subdomain routing, only path routing. If that means there are intersections you have to handle it manually in those routes.

**Reply by a-type** [reply]
For an example of how this is achieved in Next, see their [middleware](https://vercel.com/docs/concepts/functions/edge-functions#middleware) concept. Middleware is pretty much what I was looking for as I was exploring this feature. An explanation of how it's used is included in Vercel's [platform starter kit](https://vercel.com/guides/nextjs-multi-tenant-application#4.-configure-rewrites-for-multi-tenancy).

It seems to me like this should be achievable in Remix in a similar manner - either as a middleware layer which rewrites route matching information before the server renders the page, or as something higher up which changes the path passed to Remix to render (as I did with my Vercel function example above). The latter is more limited as it's platform-dependent and can't run in development mode.

Just dropping this in the discussion as a reference implementation. I think the way Next does it is quite nice as it encapsulates the entire multi-tenancy logic into middleware and doesn't leak it into the actual routing.

Edit: further investigation in Remix source seems to point to a middleware not being possible at the Remix runtime level for this. Remix processes route matching before it calls into `entry.server.ts`, so you can't intercept and rewrite the URL at that layer. Any implementation of multi-tenancy through routing middleware will have to happen at the platform layer before calling into Remix.

**Reply by prsnnami** [reply]
@a-type did you end up finding a solution for this. I was also just looking into remix for our multi-tenant app and it seems remix doesn't play well with the concept with filesystem routing and lack of a middleware.

**Reply by a-type** [reply]
Nope, for this and a few other reasons I switched back to Next, although I still think Remix does a lot of things better and easier. Hope it evolves to reach these use cases over time!

**Reply by kiliman** [reply]
Just remember that Remix does server-side routing for the initial page. Subsequent navigations are handled by React Router client-side. So URL rewriting on the server is not possible in Remix, as the URL in the browser must match the routes being processed on the client.

So your loaders will have to check the host header to see what content/layout to render.

**Reply by shawnmclean** [reply]
Damn, seems I have to stick to nextjs then, I really wanted to use remix on fly.
